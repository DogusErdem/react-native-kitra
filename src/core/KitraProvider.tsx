import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
