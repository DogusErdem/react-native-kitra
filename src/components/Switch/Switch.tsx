/* eslint-disable camelcase */
import { Switch as RNSwitch, SwitchProps as RNSwithProps } from 'react-native';
import React from 'react';
import useComponentTheme from '../../core/hooks/useComponentTheme';
import type { FCCWD, SwitchProps } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';

const Switch: FCCWD<RNSwithProps&SwitchProps> = (
  { theme, ...props },
) => {
  const { componentTheme } = useComponentTheme(theme, 'switch');

  return (
    <RNSwitch
      testID="switch"
      trackColor={{ false: componentTheme.default?.track, true: componentTheme.active?.track }}
      ios_backgroundColor={componentTheme.default?.track}
      thumbColor={props.value ? componentTheme.active?.thumb : componentTheme.default?.thumb}
      {...props}
      value={props.value}
    />
  );
};
export default applyDefaults(Switch);
