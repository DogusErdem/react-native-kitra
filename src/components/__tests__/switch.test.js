import { fireEvent, render } from '@testing-library/react-native';
import { KitraProvider } from '../../core/KitraProvider';
import Switch from '../Switch/Switch';

/* eslint-disable no-undef */
const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Switch', () => {
  it('default render', () => {
    const tree = renderWithContext(
      <Switch />,
    );
    expect(tree).toMatchSnapshot();
  });
  describe('render with onValueChange prop', () => {
    const onValueChange = jest.fn();
    const { getByTestId } = renderWithContext(
      <Switch
        onValueChange={onValueChange}
      />,
    );
    const switchElement = getByTestId('switch');
    fireEvent(switchElement, 'valueChange', true);
    expect(onValueChange).toHaveBeenCalledWith(true);
    fireEvent(switchElement, 'valueChange', false);
    expect(onValueChange).toHaveBeenCalledWith(false);
  });
});

// CUSTOM THEME EKSÄ°K
