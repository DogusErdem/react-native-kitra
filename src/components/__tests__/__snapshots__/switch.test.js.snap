// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch default render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSwitch
    accessibilityRole="switch"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onTintColor="#195CEF"
    style={
      [
        {
          "height": 31,
          "width": 51,
        },
        {
          "backgroundColor": "#DDE4F3",
          "borderRadius": 16,
        },
      ]
    }
    testID="switch"
    thumbTintColor="#FFFFFF"
    tintColor="#DDE4F3"
    typography={
      {
        "body": {
          "medium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 14,
            "lineHeight": 21,
          },
          "regular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 14,
            "lineHeight": 16,
          },
          "smedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 12,
            "lineHeight": 18,
          },
          "sregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 12,
            "lineHeight": 16,
          },
          "xsmedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 10,
            "lineHeight": 16,
          },
          "xsregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 10,
            "lineHeight": 14,
          },
          "xxsmedium": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 8,
            "lineHeight": 12,
          },
          "xxsregular": {
            "fontFamily": "Poppins-Regular",
            "fontSize": 8,
            "lineHeight": 12,
          },
        },
        "heading": {
          "h1": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 30,
            "lineHeight": 45,
          },
          "h2": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 28,
            "lineHeight": 42,
          },
          "h3": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 23,
            "lineHeight": 35,
          },
          "h4": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 19,
            "lineHeight": 29,
          },
          "h5": {
            "fontFamily": "Poppins-Medium",
            "fontSize": 16,
            "lineHeight": 25,
          },
        },
      }
    }
    value={false}
  />
</View>
`;
